### PR TITLE
LibGfx/JPEG: Support up to 4 quantization tables 

### PR DESCRIPTION
**Pull request by LucasChollet**
Images are from [w3.org/MarkUp/Test/xhtml-print/20050519/tests/A_2_1-BF-01.htm](https://www.w3.org/MarkUp/Test/xhtml-print/20050519/tests/A_2_1-BF-01.htm)

WHF

Fixes both #18455 and #18456

### CODE REVIEW

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp [hunk 1]
```diff
@@ -1072,23 +1072,23 @@ static ErrorOr<void> read_start_of_frame(Stream& stream, JPEGLoadingContext& con
 
 static ErrorOr<void> read_quantization_table(Stream& stream, JPEGLoadingContext& context)
 {
-    i32 bytes_to_read = TRY(stream.read_value<BigEndian<u16>>()) - 2;
+    u16 bytes_to_read = TRY(stream.read_value<BigEndian<u16>>()) - 2;
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:1074]: Idle musings, unrelated to this patch: We should extract a `ErrorOr<u16> read_chunk_size()` function that does:

```
static ErrorOr<u16> read_effective_chunk_size(Stream& stream)
{
    // The stored chunk size includes the size of `stored_size` itself.
    u16 stored_size = TRY(stream.read_value<BigEndian<u16>>());
    if (stored_size < 2)
        return Error::from_string_literal("stored chunk size too small");
    return stored_size - 2;
}
```

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp [hunk 2]
```diff
@@ -1072,23 +1072,23 @@ static ErrorOr<void> read_start_of_frame(Stream& stream, JPEGLoadingContext& con
...
         if (table_id > 1) {
             dbgln_if(JPEG_DEBUG, "Unsupported quantization table id: {}!", table_id);
             return Error::from_string_literal("Unsupported quantization table id");
         }
-        u32* table = table_id == 0 ? context.luma_table : context.chroma_table;
+        u32* const table = table_id == 0 ? context.luma_table : context.chroma_table;
         for (int i = 0; i < 64; i++) {
             if (element_unit_hint == 0) {
-                u8 tmp = TRY(stream.read_value<u8>());
+                u8 const tmp = TRY(stream.read_value<u8>());
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:1093]: nit: This variable is in scope for all of one line. Is that really worth it? Maybe just do

```
table[zigzag_map[i]] = TRY(stream.read_value<u8>());
```

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp [hunk 3]
```diff
@@ -1082,11 +1080,14 @@ static ErrorOr<void> read_quantization_table(Stream& stream, JPEGLoadingContext&
             return Error::from_string_literal("Unsupported unit hint in quantization table");
         }
         u8 const table_id = info_byte & 0x0F;
-        if (table_id > 1) {
+
+        if (table_id > 3) {
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:1084]: Should we keep rejecting table_id > 1 for baseline jpegs? Those are only allowed to use 2 tables.

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp [hunk 4]
```diff
@@ -237,8 +237,8 @@ struct JPEGLoadingContext {
 
     State state { State::NotDecoded };
 
-    u32 luma_table[64] = { 0 };
-    u32 chroma_table[64] = { 0 };
+    HashMap<u8, Array<u16, 64>> quantization_tables {};
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:240]: Having a HashMap for indices 0-3 seems a bit overkill. Can't this be an array of arrays?

(If the motivation is if a table is accessed that's not initialized: Maybe an array of optional arrays, but I'm guessing the spec defines some initial value for the data in these arrays?)

**Review by nico — COMMENTED** [review]
\o/

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp (hunk 3 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:1084]: From `B.2.4.1 - Quantization table-specification syntax`:

table


So it looks like even baseline DCT can have more quantization tables and btw, both faulty images are SOF0

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp (hunk 4 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:240]: An HashMap is the theoretical perfect fit IMO, but obviously overkill for 4 values. We could replace it with an array and do the job manually, but I would rather go the "mathematical" way for now and replace it later when the decoder will be optimized enough to see a difference.

And no, the JPEG spec doesn't define any default value:
`If a table has never been defined for a particular destination, then when this destination is specified in a frame header, the results are unpredictable.`

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp (hunk 1 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp]: I thought of doing this multiple time but never actually took the time :sweat_smile:

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp (hunk 4 above)
Comment (nico) [reply Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:240]: IMHO, if anything has integers 0-n with smallish n as key, then an array is the perfect fit for anyone with just a modicum of "silicon sympathy" (the usual term is "mechanical sympathy", which used to be about mechanics and actual engineering and which software folks coopted. I can't find a good link to describe it, https://dzone.com/articles/mechanical-sympathy is alright I suppose).

A hash table is also just an array, except it does a bunch of math and key reordering, so it's a useless layer of indirection in this case.

The spec says "default value can be whatever", so zeroes is fine :)

In the end, up to you, but imho using hashtables when the keys are ints 0-n with smallish n is pretty gross :)

**Review by nico — APPROVED** [review]
lgtm! (but lebtm with the remaining comment addressed)

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp (hunk 4 above)
Comment (nico) [reply Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:240]: Re parens: 

Looks like it was popularized [here](https://www.se-radio.net/2014/02/episode-201-martin-thompson-on-mechanical-sympathy/) and is originally (in software) from [here](https://mechanical-sympathy.blogspot.com/2011/07/why-mechanical-sympathy.html).

Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp]: I went with an array of optional tables, so we don't lose the ability to detect missing tables

**Review by nico — APPROVED** [review]
Nice 🙂

**Pull request merged**